File: dsa-problems/Recursion/solutions/python/REC-014-target-sum-limited-negations.py
def can_achieve_target(nums: list[int], K: int, target: int) -> bool:
    n = len(nums)
    memo = {}

    def backtrack(index, current_sum, negations):
        if index == n:
            return current_sum == target
        if (index, current_sum, negations) in memo:
            return memo[(index, current_sum, negations)]
        
        if backtrack(index + 1, current_sum + nums[index], negations):
            memo[(index, current_sum, negations)] = True
            return True
        if negations < K and backtrack(index + 1, current_sum - nums[index], negations + 1):
            memo[(index, current_sum, negations)] = True
            return True
        memo[(index, current_sum, negations)] = False
        return False

    return backtrack(0, 0, 0)

def main():
    import sys
    lines = sys.stdin.read().strip().split('\n')
    if len(lines) < 2:
        return
    first_line = lines[0].split()
    n = int(first_line[0])
    K = int(first_line[1])
    target = int(first_line[2])
    nums = list(map(int, lines[1].split()))
    result = can_achieve_target(nums, K, target)
    print("YES" if result else "NO")

if __name__ == "__main__":
    main()

File: dsa-problems/Recursion/solutions/javascript/REC-014-target-sum-limited-negations.js
const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let tokens = [];
rl.on('line', (line) => { tokens.push(...line.trim().split(/\s+/)); });
rl.on('close', () => {
    if(tokens.length===0) return;
    let ptr = 0;
    const n = parseInt(tokens[ptr++]);
    const k = parseInt(tokens[ptr++]);
    const target = parseInt(tokens[ptr++]);
    
    const nums = [];
    for(let i=0; i<n; i++) nums.push(parseInt(tokens[ptr++]));
    
    const sol = new Solution();
    if(sol.countAssignments(nums, k, target)) {
        console.log("YES");
    } else {
        console.log("NO");
    }
});

class Solution {
    countAssignments(nums, k, target) {
        this.nums = nums;
        this.k = k;
        this.target = target;
        this.n = nums.length;
        this.memo = new Map();
        return this.backtrack(0, 0, 0);
    }
    
    backtrack(index, current_sum, negations) {
        if (index === this.n) return current_sum === this.target;
        
        const key = `${index},${current_sum},${negations}`;
        if (this.memo.has(key)) return this.memo.get(key);
        
        if (this.backtrack(index + 1, current_sum + this.nums[index], negations)) {
            this.memo.set(key, true);
            return true;
        }
        
        if (negations < this.k) {
            if (this.backtrack(index + 1, current_sum - this.nums[index], negations + 1)) {
                this.memo.set(key, true);
                return true;
            }
        }
        
        this.memo.set(key, false);
        return false;
    }
}
